src/Microsoft.Azure.ServiceBus/ServiceBusDiagnosticsSource.cs

```diff
@@ -0,0 +1,771 @@
+        }
+
+        #endregion
+
+
+        #region GetSessionStateAsync
+
+        internal Activity GetSessionStateStart(string sessionId)
+        {
+            return Start("GetSessionState", () => new
+                {
+                    SessionId = sessionId,
+                    Entity = this.entityPath,
+                    Endpoint = this.endpoint
+                },
+                a => a.AddTag("SessionId", sessionId));
+        }
+
+        internal void GetSessionStateStop(Activity activity, string sessionId, byte[] state, TaskStatus? status)
+        {
+            if (activity != null)
+            {
+                DiagnosticListener.StopActivity(activity, new
+                {
+                    SessionId = sessionId,
+                    Entity = this.entityPath,
+                    Endpoint = this.endpoint,
+                    Status = status ?? TaskStatus.Faulted,
+                    State = state
+                });
+            }
+        }
+
+        #endregion
+
+
+        #region SetSessionState
+
+        internal Activity SetSessionStateStart(string sessionId, byte[] state)
+        {
+            return Start("SetSessionState", () => new
+                {
+                    State = state,
+                    SessionId = sessionId,
+                    Entity = this.entityPath,
+                    Endpoint = this.endpoint
+                },
+                a => a.AddTag("SessionId", sessionId));
+        }
+
+        internal void SetSessionStateStop(Activity activity, byte[] state, string sessionId, TaskStatus? status)
+        {
+            if (activity != null)
+            {
+                DiagnosticListener.StopActivity(activity, new
+                {
+                    State = state,
+                    SessionId = sessionId,
+                    Entity = this.entityPath,
+                    Endpoint = this.endpoint,
+                    Status = status ?? TaskStatus.Faulted
+                });
+            }
+        }
+
+        #endregion
+
+
+        #region RenewSessionLock
+
+        internal Activity RenewSessionLockStart(string sessionId)
+        {
+            return Start("RenewSessionLock", () => new
+            {
+                SessionId = sessionId,
+                Entity = this.entityPath,
+                Endpoint = this.endpoint
+            },
+            a => a.AddTag("SessionId", sessionId));
+        }
+
+        internal void RenewSessionLockStop(Activity activity, string sessionId, TaskStatus? status)
+        {
+            if (activity != null)
+            {
+                DiagnosticListener.StopActivity(activity, new
+                {
+                    SessionId = sessionId,
+                    Entity = this.entityPath,
+                    Endpoint = this.endpoint,
+                    Status = status ?? TaskStatus.Faulted
+                });
+            }
+        }
+
+        #endregion
+
+        internal void ReportException(Exception ex)
```

Comment:
>internal void ReportException(Exception ex) [](start = 8, length = 43)

Is there a paradigm of `ReportException` being associated with an `Activity`?

Reply:
@lmolkova  What I meant was, we are just reporting exception here without associating it with an activity. Shouldn't we be passing an activity?
We associate all starts and stops of operations across machines. Will we be able to associate the exception with a correlationId?

Reply:
Activity has static Current property that flows with the async calls, i.e. users can always use Activity.Current, it does not have to be explicitly passed to the event

Reply:
e.g. in exception event handler you can write
```C#
Logger.LogError($"Exception {ex}, Id: {Activity.Current?.Id}")
```